nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/record/AbstractKinesisRecordProcessor.java

```diff
@@ -0,0 +1,281 @@
+
+    final ProcessSessionFactory sessionFactory;
+    final ComponentLog log;
+    final String streamName;
+    final String endpointPrefix;
+
+    final long checkpointIntervalMillis;
+    final long retryWaitMillis;
+    final int numRetries;
+    final DateTimeFormatter dateTimeFormatter;
+
+    String kinesisShardId;
+    long nextCheckpointTimeInMillis;
+
+    boolean processingRecords = false;
+
+    @SuppressWarnings("java:S107")
+    AbstractKinesisRecordProcessor(final ProcessSessionFactory sessionFactory, final ComponentLog log, final String streamName,
+                                   final String endpointPrefix, final long checkpointIntervalMillis, final long retryWaitMillis,
+                                   final int numRetries, final DateTimeFormatter dateTimeFormatter) {
+        this.sessionFactory = sessionFactory;
+        this.log = log;
+        this.streamName = streamName;
+        this.endpointPrefix = endpointPrefix;
+        this.checkpointIntervalMillis = checkpointIntervalMillis;
+        this.retryWaitMillis = retryWaitMillis;
+        this.numRetries = numRetries;
+        this.dateTimeFormatter = dateTimeFormatter;
+    }
+
+    @Override
+    public void initialize(final InitializationInput initializationInput) {
+        if (initializationInput.getPendingCheckpointSequenceNumber() != null) {
+            log.warn("Initializing record processor for stream: {} / shard {}; from sequence number: {}; indicates previously uncheckpointed sequence number: {}",
+                    streamName, initializationInput.getShardId(), initializationInput.getExtendedSequenceNumber(), initializationInput.getPendingCheckpointSequenceNumber());
+        } else {
+            log.debug("Initializing record processor for stream: {} / shard: {}; from sequence number: {}",
+                    streamName, initializationInput.getShardId(), initializationInput.getExtendedSequenceNumber());
+        }
+
+        this.kinesisShardId = initializationInput.getShardId();
+
+        // ensure we don't immediately checkpoint
+        this.nextCheckpointTimeInMillis = System.currentTimeMillis() + checkpointIntervalMillis;
+    }
+
+    @SuppressWarnings({"java:S3252"}) // ConsumeKinesisStream reference to REL_SUCCESS instead of deprecated AbstractAWSProcessor
+    @Override
+    public void processRecords(final ProcessRecordsInput processRecordsInput) {
+        log.debug("Processing {} records from {}; cache entry: {}; cache exit: {}; millis behind latest: {}",
+                processRecordsInput.getRecords().size(), kinesisShardId,
+                processRecordsInput.getCacheEntryTime() != null ? dateTimeFormatter.format(processRecordsInput.getCacheEntryTime().atZone(ZoneId.systemDefault())) : null,
+                processRecordsInput.getCacheExitTime() != null ? dateTimeFormatter.format(processRecordsInput.getCacheExitTime().atZone(ZoneId.systemDefault())) : null,
+                processRecordsInput.getMillisBehindLatest());
```

Comment:
I don't really like `if (log.isDebugEnabled()) { ... }` but in this case I would add it because `processRecords()` gets called frequently and there are date conversions in the debug call.

Reply:
Tried to catch these, but evidently missed one... good spot